src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
-    select_i[0] = 0;
-    int cdof_counter = 0;
-    for (int i = 0; i < nfaces; i++)
+
+        h_D_trace.MultAT(h_F_potential, DTT_F);
+        h_D_trace.MultAT(B_potential_[agg], DTT_B);
+
+        M_el[agg].SetSubMatrix(num_bubbles, num_bubbles, DTT_F);
+        M_el[agg].SetSubMatrix(num_bubbles, 0, DTT_B);
+        M_el[agg].SetSubMatrixTranspose(0, num_bubbles, DTT_B);
+    }
+
+    Vector M_local;
+    DenseMatrix edge_target_T_M;
+    DenseMatrix trace_across;
+
+    std::vector<int> agg_indices;
+
+    SparseMatrix edge_vertex = mgl.LocalD().Transpose();
+    SparseMatrix vertex_agg = gt_.agg_vertex_local_.Transpose();
+
+    for (int face = 0; face < num_faces; ++face)
     {
-        face_ncdofs = face_cdof_i[i + 1] - face_cdof_i[i];
-        if (face_d_td_diag.RowSize(i))
-            for (int j = 0; j < face_ncdofs; j++)
-            {
-                select_i[cdof_counter + 1] = select_i[cdof_counter] + 1;
-                cdof_counter++;
-            }
-        else
-            for (int j = 0; j < face_ncdofs; j++)
+        std::vector<int> fine_dofs = gt_.face_edge_local_.GetIndices(face);
+        std::vector<int> coarse_dofs = face_cdof_.GetIndices(face);
+        std::vector<int> aggs = gt_.face_agg_local_.GetIndices(face);
+        agg_indices.resize(coarse_dofs.size());
+
+        for (auto agg : aggs)
+        {
+            BuildAggregateFaceM(mgl, agg, fine_dofs, vertex_agg,
+                                edge_vertex, col_marker_, M_local);
+
+            edge_targets_[face].Transpose(edge_target_T_M);
+            edge_target_T_M.ScaleCols(M_local);
+
+            edge_target_T_M.Mult(edge_targets_[face], trace_across);
+
+            std::vector<int> edges = agg_cdof_edge.GetIndices(agg);
+            SetMarker(col_marker_, edges);
+
+            int size = coarse_dofs.size();
+            for (int i = 0; i < size; ++i)
             {
-                select_i[cdof_counter + 1] = select_i[cdof_counter];
-                cdof_counter++;
+                agg_indices[i] = col_marker_[coarse_dofs[i]];
+                assert(agg_indices[i] > -1);
             }
+
+            M_el[agg].AddSubMatrix(agg_indices, agg_indices, trace_across);
+
+            ClearMarker(col_marker_, edges);
+        }
+    }
+
+    return M_el;
+}
+
+DenseMatrix GraphCoarsen::RestrictLocal(const DenseMatrix& ext_mat,
+                                        std::vector<int>& global_marker,
+                                        const std::vector<int>& ext_indices,
+                                        const std::vector<int>& local_indices) const
+{
+    SetMarker(global_marker, ext_indices);
+
+    int local_size = local_indices.size();
+
+    std::vector<int> row_map(local_size);
+
+    for (int i = 0; i < local_size; ++i)
+    {
+        assert(global_marker[local_indices[i]] >= 0);
+        row_map[i] = global_marker[local_indices[i]];
+    }
+
+    ClearMarker(global_marker, ext_indices);
+
+    return ext_mat.GetRow(row_map);
```

Comment:
(This should be pretty minor)

The original implementation is to first take transpose, extract columns, and then transpose the columns, which is copied from `ParElag`.

I remember at some point I was curious why not directly extract rows (what you are doing here), so I compare the two approaches. I found that the `ParElag` approach is actually faster, albeit it does extra transpose action twice. It was out of my expectation, and I think it's probably due to the storage format of DenseMatrix (data are concatenated columns) that makes copying columns (using `std::copy`) much faster. Did you observe similar behavior? 

`DenseMatrix::GetCol` in `linalgcpp` is using `GetSubMatrix` though, I think it may worth to use `std::copy` for columns, especially for continuous column indices.